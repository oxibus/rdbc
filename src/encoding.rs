// https://github.com/hsivonen/recode_rs

use std::io::{Read, Write};
use std::str::from_utf8_mut;

use encoding_rs::*;

use crate::error::DbcError;

pub fn utf8_to_gbk(src_data: &[u8]) -> Result<Vec<u8>, DbcError> {
    recode(src_data, "UTF-8", "GBK")
}

pub fn gbk_to_utf8(src_data: &[u8]) -> Result<Vec<u8>, DbcError> {
    recode(src_data, "GBK", "UTF-8")
}

pub fn to_utf8(src_encoding_label: &str, src_data: &[u8]) -> Result<Vec<u8>, DbcError> {
    recode(src_data, src_encoding_label, "UTF-8")
}

pub fn recode(
    src_data: &[u8],
    src_encoding_label: &str,
    dst_encoding_label: &str,
) -> Result<Vec<u8>, DbcError> {
    let src_encoding = get_encoding(Some(src_encoding_label.to_string()))?;
    let dst_encoding = get_encoding(Some(dst_encoding_label.to_string()))?;

    let mut decoder = src_encoding.new_decoder();
    let mut encoder = dst_encoding.new_encoder();

    let mut buf = std::io::Cursor::new(Vec::new());

    convert_via_utf8(
        &mut decoder,
        &mut encoder,
        &mut std::io::Cursor::new(src_data),
        &mut buf,
        true,
    )?;

    Ok(buf.into_inner())
}

pub fn get_encoding(opt: Option<String>) -> Result<&'static Encoding, DbcError> {
    match opt {
        None => Ok(UTF_8),
        Some(label) => match Encoding::for_label(label.as_bytes()) {
            None => Err(DbcError::InvalidEncodingLabel(label)),
            Some(encoding) => Ok(encoding),
        },
    }
}

pub fn convert_via_utf8(
    decoder: &mut Decoder,
    encoder: &mut Encoder,
    read: &mut dyn Read,
    write: &mut dyn Write,
    last: bool,
) -> Result<(), DbcError> {
    let mut input_buffer = [0u8; 2048];
    let mut intermediate_buffer_bytes = [0u8; 4096];
    let intermediate_buffer: &mut str = from_utf8_mut(&mut intermediate_buffer_bytes[..]).unwrap();
    let mut output_buffer = [0u8; 4096];
    let mut current_input_ended = false;
    while !current_input_ended {
        match read.read(&mut input_buffer) {
            Err(e) => {
                log::error!("Error reading input, error = {}", e);
                return Err(DbcError::EncodingReadInputError);
            }
            Ok(decoder_input_end) => {
                current_input_ended = decoder_input_end == 0;
                let input_ended = last && current_input_ended;
                let mut decoder_input_start = 0usize;
                loop {
                    let (decoder_result, decoder_read, decoder_written, _) = decoder.decode_to_str(
                        &input_buffer[decoder_input_start..decoder_input_end],
                        intermediate_buffer,
                        input_ended,
                    );
                    decoder_input_start += decoder_read;

                    let last_output = if input_ended {
                        match decoder_result {
                            CoderResult::InputEmpty => true,
                            CoderResult::OutputFull => false,
                        }
                    } else {
                        false
                    };

                    // Regardless of whether the intermediate buffer got full
                    // or the input buffer was exhausted, let's process what's
                    // in the intermediate buffer.

                    if encoder.encoding() == UTF_8 {
                        // If the target is UTF-8, optimize out the encoder.
                        if let Err(e) =
                            write.write_all(&intermediate_buffer.as_bytes()[..decoder_written])
                        {
                            log::error!("Error writing output, error = {}", e);
                            return Err(DbcError::EncodingWriteOutputError);
                        }
                    } else {
                        let mut encoder_input_start = 0usize;
                        loop {
                            let (encoder_result, encoder_read, encoder_written, _) = encoder
                                .encode_from_utf8(
                                    &intermediate_buffer[encoder_input_start..decoder_written],
                                    &mut output_buffer,
                                    last_output,
                                );
                            encoder_input_start += encoder_read;
                            if let Err(e) = write.write_all(&output_buffer[..encoder_written]) {
                                log::error!("Error writing output, error = {}", e);
                                return Err(DbcError::EncodingWriteOutputError);
                            }
                            match encoder_result {
                                CoderResult::InputEmpty => {
                                    break;
                                }
                                CoderResult::OutputFull => {
                                    continue;
                                }
                            }
                        }
                    }

                    // Now let's see if we should read again or process the
                    // rest of the current input buffer.
                    match decoder_result {
                        CoderResult::InputEmpty => {
                            break;
                        }
                        CoderResult::OutputFull => {
                            continue;
                        }
                    }
                }
            }
        }
    }

    Ok(())
}
