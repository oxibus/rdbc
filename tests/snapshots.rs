#![cfg(feature = "serde")]

use std::fs;
use std::path::Path;

use insta::{assert_debug_snapshot, assert_snapshot, assert_yaml_snapshot, with_settings};
use rrdbc::ast::network_ast::parse_dbc;
use rrdbc::encoding::decode_cp1252;

/// Test parsing all DBC files
#[test]
fn test_cantools_dbc_files() {
    let test_dirs = [
        ("dbc/dbc_from_cantools", "dbc-cantools"),
        ("dbc/mytest", "mytest"),
    ];
    for (path, file_name) in &test_dirs {
        let path = Path::new(path);
        let snapshot_path = String::from("snapshots-") + file_name;
        // snapshots should go to /tests/snapshots-* directory
        with_settings! {
            { omit_expression => true,
              snapshot_path => snapshot_path ,
              prepend_module_to_snapshot => false },
            {
                test_dbc_files(path);
            }
        }
    }
}

/// Test parsing all DBC files in the given directory.
fn test_dbc_files(dir: impl AsRef<Path>) {
    let dir = dir.as_ref();
    let dir_display = dir.display();
    let dir_content = fs::read_dir(dir).unwrap_or_else(|e| {
        panic!(
            "
--------------------------------------------------------------------------
Error reading dbc test files from   {dir_display}
{e}
Make sure git submodules are up to date by running
    git submodule update --init --recursive
--------------------------------------------------------------------------
"
        )
    });

    eprintln!("Testing dbc files in directory: {dir_display}");
    for dbc_path in dir_content {
        let path = dbc_path.unwrap().path();
        if path
            .extension()
            .is_some_and(|ext| ext.eq_ignore_ascii_case("dbc"))
        {
            parse_one_file(&path);
        }
    }
}

/// Parse a single DBC file and assert a snapshot of the result.
fn parse_one_file(path: &Path) {
    eprintln!("Testing DBC file: {}", path.display());
    let file_name = path.file_stem().unwrap().to_string_lossy().to_string();
    let buffer = fs::read(path).unwrap();
    if let Some(buffer) = decode_cp1252(&buffer) {
        match parse_dbc(buffer.as_ref()) {
            Ok(dbc) => assert_yaml_snapshot!(file_name, dbc),
            Err(e) => {
                eprintln!("Failed to parse {file_name}.dbc: {e:?}");
                assert_debug_snapshot!(format!("!error___{file_name}"), e);
            }
        }
    } else {
        let error = format!("Failed to decode {file_name}.dbc as cp1252");
        eprintln!("{error}");
        assert_snapshot!(format!("!error___{file_name}"), error);
    }
}
