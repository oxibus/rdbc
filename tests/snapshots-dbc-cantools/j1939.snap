---
source: tests/snapshots.rs
---
version: ""
new_symbols:
  - NS_DESC_
  - CM_
  - BA_DEF_
  - BA_
  - VAL_
  - CAT_DEF_
  - CAT_
  - FILTER
  - BA_DEF_DEF_
  - EV_DATA_
  - ENVVAR_DATA_
  - SGTYPE_
  - SGTYPE_VAL_
  - BA_DEF_SGTYPE_
  - BA_SGTYPE_
  - SIG_TYPE_REF_
  - VAL_TABLE_
  - SIG_GROUP_
  - SIG_VALTYPE_
  - SIGTYPE_VALTYPE_
  - BO_TX_BU_
  - BA_DEF_REL_
  - BA_REL_
  - BA_DEF_DEF_REL_
  - BU_SG_REL_
  - BU_EV_REL_
  - BU_BO_REL_
  - SG_MUL_VAL_
bit_timing:
  value: ~
nodes:
  - Node2
  - Node1
value_tables: ~
messages:
  - header:
      id: 2503213569
      name: Message1
      size: 8
      transmitter: Node1
    signals:
      - name: Signal1
        multiplexer: ~
        start_bit: 0
        size: 8
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Node2
  - header:
      id: 2515537922
      name: Message2
      size: 8
      transmitter: Node2
    signals:
      - name: Signal2
        multiplexer: ~
        start_bit: 0
        size: 8
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Node1
env_vars: []
env_vars_data: []
comments: []
attribute_definitions:
  - Message:
      attribute_name: TpJ1939VarDlc
      attribute_value_type:
        Enum:
          values:
            - "No"
            - "Yes"
  - Signal:
      attribute_name: SigType
      attribute_value_type:
        Enum:
          values:
            - Default
            - Range
            - RangeSigned
            - ASCII
            - Discrete
            - Control
            - ReferencePGN
            - DTC
            - StringDelimiter
            - StringLength
            - StringLengthControl
            - MessageCounter
            - MessageChecksum
  - Signal:
      attribute_name: GenSigEVName
      attribute_value_type:
        String: {}
  - Signal:
      attribute_name: GenSigILSupport
      attribute_value_type:
        Enum:
          values:
            - "No"
            - "Yes"
  - Signal:
      attribute_name: GenSigSendType
      attribute_value_type:
        Enum:
          values:
            - Cyclic
            - OnWrite
            - OnWriteWithRepetition
            - OnChange
            - OnChangeWithRepetition
            - IfActive
            - IfActiveWithRepetition
            - NoSigSendType
  - Message:
      attribute_name: GenMsgFastOnStart
      attribute_value_type:
        Integer:
          minimum: 0
          maximum: 100000
  - Signal:
      attribute_name: GenSigInactiveValue
      attribute_value_type:
        Integer:
          minimum: 0
          maximum: 0
  - Message:
      attribute_name: GenMsgCycleTimeFast
      attribute_value_type:
        Integer:
          minimum: 0
          maximum: 3600000
  - Message:
      attribute_name: GenMsgNrOfRepetition
      attribute_value_type:
        Integer:
          minimum: 0
          maximum: 1000000
  - Signal:
      attribute_name: GenSigStartValue
      attribute_value_type:
        Integer:
          minimum: 0
          maximum: 2147483647
  - Message:
      attribute_name: GenMsgDelayTime
      attribute_value_type:
        Integer:
          minimum: 0
          maximum: 1000
  - Message:
      attribute_name: GenMsgILSupport
      attribute_value_type:
        Enum:
          values:
            - "No"
            - "Yes"
  - Message:
      attribute_name: GenMsgStartDelayTime
      attribute_value_type:
        Integer:
          minimum: 0
          maximum: 100000
  - Node:
      attribute_name: NodeLayerModules
      attribute_value_type:
        String: {}
  - Node:
      attribute_name: ECU
      attribute_value_type:
        String: {}
  - Node:
      attribute_name: NmJ1939SystemInstance
      attribute_value_type:
        Integer:
          minimum: 0
          maximum: 15
  - Node:
      attribute_name: NmJ1939System
      attribute_value_type:
        Integer:
          minimum: 0
          maximum: 127
  - Node:
      attribute_name: NmJ1939ManufacturerCode
      attribute_value_type:
        Integer:
          minimum: 0
          maximum: 2047
  - Node:
      attribute_name: NmJ1939IndustryGroup
      attribute_value_type:
        Integer:
          minimum: 0
          maximum: 7
  - Node:
      attribute_name: NmJ1939IdentityNumber
      attribute_value_type:
        Integer:
          minimum: 0
          maximum: 2097151
  - Node:
      attribute_name: NmJ1939FunctionInstance
      attribute_value_type:
        Integer:
          minimum: 0
          maximum: 7
  - Node:
      attribute_name: NmJ1939Function
      attribute_value_type:
        Integer:
          minimum: 0
          maximum: 255
  - Node:
      attribute_name: NmJ1939ECUInstance
      attribute_value_type:
        Integer:
          minimum: 0
          maximum: 3
  - Node:
      attribute_name: NmJ1939AAC
      attribute_value_type:
        Integer:
          minimum: 0
          maximum: 1
  - Node:
      attribute_name: NmStationAddress
      attribute_value_type:
        Integer:
          minimum: 0
          maximum: 255
  - Message:
      attribute_name: GenMsgSendType
      attribute_value_type:
        Enum:
          values:
            - cyclic
            - NotUsed
            - IfActive
            - NotUsed
            - NotUsed
            - NotUsed
            - NotUsed
            - NotUsed
            - noMsgSendType
  - Message:
      attribute_name: GenMsgRequestable
      attribute_value_type:
        Integer:
          minimum: 0
          maximum: 1
  - Message:
      attribute_name: GenMsgCycleTime
      attribute_value_type:
        Integer:
          minimum: 0
          maximum: 3600000
  - Signal:
      attribute_name: SPN
      attribute_value_type:
        Integer:
          minimum: 0
          maximum: 524287
  - Network:
      attribute_name: DBName
      attribute_value_type:
        String: {}
  - Network:
      attribute_name: BusType
      attribute_value_type:
        String: {}
  - Network:
      attribute_name: ProtocolType
      attribute_value_type:
        String: {}
  - Message:
      attribute_name: VFrameFormat
      attribute_value_type:
        Enum:
          values:
            - StandardCAN
            - ExtendedCAN
            - reserved
            - J1939PG
attribute_defaults:
  - Attribute:
      attribute_name: TpJ1939VarDlc
      attribute_value:
        String: "No"
  - Attribute:
      attribute_name: SigType
      attribute_value:
        String: Default
  - Attribute:
      attribute_name: GenSigEVName
      attribute_value:
        String: Env@Nodename_@Signame
  - Attribute:
      attribute_name: GenSigILSupport
      attribute_value:
        String: "Yes"
  - Attribute:
      attribute_name: GenSigSendType
      attribute_value:
        String: NoSigSendType
  - Attribute:
      attribute_name: GenMsgFastOnStart
      attribute_value:
        Double: 0
  - Attribute:
      attribute_name: GenSigInactiveValue
      attribute_value:
        Double: 0
  - Attribute:
      attribute_name: GenMsgCycleTimeFast
      attribute_value:
        Double: 0
  - Attribute:
      attribute_name: GenMsgNrOfRepetition
      attribute_value:
        Double: 0
  - Attribute:
      attribute_name: GenSigStartValue
      attribute_value:
        Double: 0
  - Attribute:
      attribute_name: GenMsgDelayTime
      attribute_value:
        Double: 0
  - Attribute:
      attribute_name: GenMsgILSupport
      attribute_value:
        String: "Yes"
  - Attribute:
      attribute_name: GenMsgStartDelayTime
      attribute_value:
        Double: 0
  - Attribute:
      attribute_name: NodeLayerModules
      attribute_value:
        String: ""
  - Attribute:
      attribute_name: ECU
      attribute_value:
        String: ""
  - Attribute:
      attribute_name: NmJ1939SystemInstance
      attribute_value:
        Double: 0
  - Attribute:
      attribute_name: NmJ1939System
      attribute_value:
        Double: 0
  - Attribute:
      attribute_name: NmJ1939ManufacturerCode
      attribute_value:
        Double: 0
  - Attribute:
      attribute_name: NmJ1939IndustryGroup
      attribute_value:
        Double: 0
  - Attribute:
      attribute_name: NmJ1939IdentityNumber
      attribute_value:
        Double: 0
  - Attribute:
      attribute_name: NmJ1939FunctionInstance
      attribute_value:
        Double: 0
  - Attribute:
      attribute_name: NmJ1939Function
      attribute_value:
        Double: 0
  - Attribute:
      attribute_name: NmJ1939ECUInstance
      attribute_value:
        Double: 0
  - Attribute:
      attribute_name: NmJ1939AAC
      attribute_value:
        Double: 0
  - Attribute:
      attribute_name: NmStationAddress
      attribute_value:
        Double: 254
  - Attribute:
      attribute_name: GenMsgSendType
      attribute_value:
        String: noMsgSendType
  - Attribute:
      attribute_name: GenMsgRequestable
      attribute_value:
        Double: 1
  - Attribute:
      attribute_name: GenMsgCycleTime
      attribute_value:
        Double: 0
  - Attribute:
      attribute_name: SPN
      attribute_value:
        Double: 0
  - Attribute:
      attribute_name: DBName
      attribute_value:
        String: ""
  - Attribute:
      attribute_name: BusType
      attribute_value:
        String: CAN
  - Attribute:
      attribute_name: ProtocolType
      attribute_value:
        String: J1939
  - Attribute:
      attribute_name: VFrameFormat
      attribute_value:
        String: J1939PG
attribute_values:
  - Network:
      attribute_name: DBName
      attribute_value:
        String: foo
  - Node:
      attribute_name: NmStationAddress
      node_name: Node2
      attribute_value:
        Double: 2
  - Node:
      attribute_name: NmStationAddress
      node_name: Node1
      attribute_value:
        Double: 1
  - Message:
      attribute_name: VFrameFormat
      message_id: 2503213569
      attribute_value:
        Double: 3
  - Signal:
      attribute_name: SPN
      message_id: 2503213569
      signal_name: Signal1
      attribute_value:
        Double: 500
signal_value_descriptions: []
env_var_value_descriptions: []
